cpp/serve/engine_actions/batch_draft.cc

```diff
@@ -0,0 +1,213 @@
+ * `running_queue` of engine state. Preempt low-priority requests
+ * accordingly when it is impossible to decode all the running requests.
+ */
+class BatchDraftActionObj : public EngineActionObj {
+ public:
+  explicit BatchDraftActionObj(Array<Model> models, Sampler sampler,
+                               Optional<EventTraceRecorder> trace_recorder, int draft_length = 4)
+      : models_(std::move(models)),
+        sampler_(std::move(sampler)),
+        trace_recorder_(std::move(trace_recorder)),
+        draft_length_(draft_length) {}
+
+  Array<Request> Step(EngineState estate) final {
+    // - Only run spec decode when there are two models (llm+ssm) and >=1 running requests.
+    if (models_.size() != 2 || estate->running_queue.empty()) {
+      return {};
+    }
+
+    // Preempt requests when decode cannot apply.
+    while (!CanDecode(estate->running_queue.size())) {
+      PreemptLastRunningRequest(estate);
+    }
+
+    auto tstart = std::chrono::high_resolution_clock::now();
+
+    // NOTE: Right now we only support decode all the running requests at a time.
+    int num_requests = estate->running_queue.size();
+
+    // The first model doesn't get involved in draft proposal.
+    for (int model_id = 1; model_id < models_.size(); ++model_id) {
+      // draft_length_ rounds of draft proposal.
+      for (int draft_id = 0; draft_id < draft_length_; ++draft_id) {
+        // Collect
+        // - the last committed token,
+        // - the request states,
+        // - the sampling parameters,
+        // of each request.
+        std::vector<int> input_tokens;
+        Array<String> request_ids;
+        std::vector<int64_t> request_internal_ids;
+        Array<RequestModelState> mstates;
+        Array<GenerationConfig> generation_cfg;
+        input_tokens.reserve(num_requests);
+        request_ids.reserve(num_requests);
+        request_internal_ids.reserve(num_requests);
+        mstates.reserve(num_requests);
+        generation_cfg.reserve(num_requests);
+        for (Request request : estate->running_queue) {
+          RequestState rstate = estate->GetRequestState(request);
+          if (draft_id == 0) {
+            // The first draft proposal uses the last committed token.
+            input_tokens.push_back(rstate->mstates[model_id]->committed_tokens.back());
+          } else {
+            input_tokens.push_back(rstate->mstates[model_id]->draft_output_tokens.back());
+          }
+          request_ids.push_back(request->id);
+          request_internal_ids.push_back(rstate->mstates[0]->internal_id);
+          mstates.push_back(rstate->mstates[model_id]);
+          generation_cfg.push_back(request->generation_cfg);
+        }
+
+        // - Compute embeddings.
+        RECORD_EVENT(trace_recorder_, request_ids,
+                     "start propsal embedding for model_id " + std::to_string(model_id) +
+                         " draft_id " + std::to_string(draft_id));
```

Comment:
We can make all the message lightweight.
```suggestion
        RECORD_EVENT(trace_recorder_, request_ids, "start proposal embedding");
```